### ISSUE TITLE
Type refinment with optional chaining

### ISSUE THREAD

**Issue by antonk52** [microsoft/TypeScript#40397]
**TypeScript Version:**  v4.0.2

**Search Terms:** optional chaining, type refinement

**Code**

```ts
type A = {
  type: 'A';
  name: string;
}

type B = {
  type: 'B';
}

// example one
function funcOne (arg: A | B | undefined) {
  const errorOne = arg?.name; // unexpected error ❌

  errorOne; // expected `string | undefined` but got `any`

  return errorOne;
}

// example two
function funcTwo (arg: A | B | undefined) {
  // note the optional chaining in the condition
  if (arg?.type === 'B') {
    arg; // `B`
    return 'B';
  }

  arg; // `A | B | undefined` ❌

  // arg cannot be of type B here
  const errorTwo = arg?.name; // unexpected error ❌

  return errorTwo;
}

// same as example 2, but no optional chaining in the condition
function funcTwoButWorks (arg: A | B | undefined) {
  // no optional chaining
  if (arg !== undefined && arg.type === 'B') {
    arg; // `B`
    return 'B';
  } 

  arg; // `A | undefined` ✅

  const result = arg?.name; // no error, yay ✅

  return result;
}
```

**Expected behavior:**
- in `funcOne` there should be no error and the result should be of type `string | undefined`.
- in `funcTwo` after the the if condition check `arg` should be of type `A | undefined`.

**Actual behavior:**
- in `funcOne` errors for optional chaining instead of inferring the type as `string | undefined`
- in `funcTwo` it does not extract type B from possible values for `arg`

**Playground Link:** [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeAoKVSQFxQOQ1wG40oA7AQwFsIcBnYAJwEtSBzYgXxRU2gCEEyErxy4+RFFxQB6aVAgAPKmAA20APakIKAGYBXUgGNgTTVH1GA8lqgAKcg1Y44AHygC3BgCYQdLCF4AlELohpr08gwM6gzW0IgOrAD8AHQU1IRQslAGipDGAZHRDFCAMuTc6BBRMXGZ2XkQBV5QAAb0zGxQnqQ+flpeLVAARnrAUKzqYy3kpCAtFVAMEMB6DKRFNVqc3PVKlKrQwADu6roGxqbrFoYAKid2ic5d7s-evv5BIVlypJOHABYaMAmTTkFRQQz-cgsFisKAsDCAiGaLxMEGkEhMHQPRypXgIeCIMS4YKodDoRJ1OQtPjzcmLZardbE4joKQUxxU1quF7dXofQblEjZRIQma-MZDDTY-ECQFLEhhUgRKrFO7qQSJVLpCBc3IKfLAQqqmJlBZLFZrDYMdXbGRyWhUaDkWjyPYHKAAJgANMNRmQNepgZcwRCoTDOgjgEilaj0WcjOjzOd1XxRgB1GIAa1d9kcTzcHhyPXe-VJwp+geDoPBkOhpFhmOxebhAEJCcX+f0oAAyHtQRIpfGEonicv0ynfVq0kjoC1MvDiVlQDhQBaT7ItHlvPoBQaAUHIFkqIktaHoVGMEri0k6ub9rb6QOQQFBDyR51bT+fgJwgA) 

**Related Issues:** maybe [38136](https://github.com/microsoft/TypeScript/issues/38136).

**Comment by MartinJohns** [comment]
Example one works as expected - you can only access properties shared across all types of a union. What you want would be unsafe, because TypeScript has a structural type system. The value for `B` could be of the type `{ type: "B", name: number }`, so inferring `string | undefined` would be very bad.

**Comment by whzx5byb** [comment]
If you want to get type `string | undefined` from `arg?.name` in `funcOne`, you should make the type B like this:
```
type B = {
  type: 'B';
  name?: never;
}
```

**Comment by andrewbranch** [comment]
This is fixed by #38839. Not sure how much of that PR is necessary to fix this, which seems like a fairly focused problem.
